pkg/storegateway/bucket.go

```diff
@@ -1811,35 +1751,52 @@ func (s *bucketBlockSet) add(b *bucketBlock) error {
 		}
 		return s.blocks[j].meta.MinTime < s.blocks[k].meta.MinTime
 	})
-	return nil
 }
 
-func (s *bucketBlockSet) remove(id ulid.ULID) {
+// remove removes the block identified by id from the set. It returns the removed block if it was present in the set.
+func (s *bucketBlockSet) remove(id ulid.ULID) *bucketBlock {
 	s.mtx.Lock()
 	defer s.mtx.Unlock()
 
+	val, ok := s.blockSet.LoadAndDelete(id)
+	if !ok {
+		return nil
+	}
+
 	for i, b := range s.blocks {
 		if b.meta.ULID != id {
 			continue
 		}
 		s.blocks = append(s.blocks[:i], s.blocks[i+1:]...)
-		return
+		break
 	}
+
+	return val.(*bucketBlock)
 }
 
-// getFor returns a time-ordered list of blocks that cover date between mint and maxt.
-// It supports overlapping blocks.
-//
-// NOTE: s.blocks are expected to be sorted in minTime order.
-func (s *bucketBlockSet) getFor(mint, maxt int64, blockMatchers []*labels.Matcher) (bs []*bucketBlock) {
+func (s *bucketBlockSet) has(id ulid.ULID) bool {
+	_, ok := s.blockSet.Load(id)
+	return ok
+}
+
+func (s *bucketBlockSet) len() int {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+	return len(s.blocks)
+}
+
+// filter iterates over a time-ordered list of non-closed blocks that cover date between mint and maxt. It supports overlapping
+// blocks. A block is guarantied to be not closed only inside the fn.
```

Comment:
well technically `inside fn and if fn invoked a function on the block which added to pendingReaders`, but this sounds clunky. Maybe we should rephrase to `A block is guaranteed to not be closed while fn is executing`.